pkg/reconciler/broker/resources/names.go

```diff
@@ -30,34 +43,60 @@ const (
 	md5Len          = 32
 )
 
-//TODO check length < 256 chars and truncate
-// pubsub resource max length: 255 chars
-// Namespace max length: 63 chars
-// broker name max length: 253 chars
-// trigger name max length: 253 chars
-// uid length: 36 chars
-// prefix + separators: 10 chars
-// 255 - 10 - 36 - 63 = 146 chars for name
-
 // GenerateDecouplingTopicName generates a deterministic topic name for a
-// Broker.
+// Broker. If the topic name would be longer than allowed by PubSub, the
+// Broker name is truncated to fit.
 func GenerateDecouplingTopicName(b *brokerv1beta1.Broker) string {
-	return fmt.Sprintf("cre-bkr_%s_%s_%s", b.Namespace, b.Name, string(b.UID))
+	return truncatedPubsubResourceName("cre-bkr", b.Namespace, b.Name, string(b.UID))
 }
 
 // GenerateDecouplingSubscriptionName generates a deterministic subscription
-// name for a Broker.
+// name for a Broker. If the subscription name would be longer than allowed by
+// PubSub, the Broker name is truncated to fit.
 func GenerateDecouplingSubscriptionName(b *brokerv1beta1.Broker) string {
-	return fmt.Sprintf("cre-bkr_%s_%s_%s", b.Namespace, b.Name, string(b.UID))
+	return truncatedPubsubResourceName("cre-bkr", b.Namespace, b.Name, string(b.UID))
 }
 
 // GenerateRetryTopicName generates a deterministic topic name for a Trigger.
+// If the topic name would be longer than allowed by PubSub, the Trigger name is
+// truncated to fit.
 func GenerateRetryTopicName(t *brokerv1beta1.Trigger) string {
-	return fmt.Sprintf("cre-tgr_%s_%s_%s", t.Namespace, t.Name, string(t.UID))
+	return truncatedPubsubResourceName("cre-tgr", t.Namespace, t.Name, string(t.UID))
 }
 
 // GenerateRetrySubscriptionName generates a deterministic subscription name
-// for a Trigger.
+// for a Trigger. If the subscription name would be longer than allowed by
+// PubSub, the Trigger name is truncated to fit.
 func GenerateRetrySubscriptionName(t *brokerv1beta1.Trigger) string {
-	return fmt.Sprintf("cre-tgr_%s_%s_%s", t.Namespace, t.Name, string(t.UID))
+	return truncatedPubsubResourceName("cre-tgr", t.Namespace, t.Name, string(t.UID))
+}
+
+func truncatedPubsubResourceName(prefix, ns, n, uid string) string {
+	s := fmt.Sprintf("%s_%s_%s_%s", prefix, ns, n, uid)
+	if len(s) <= pubsubMax {
+		return s
+	}
+	// Calculate how much the name needs to be truncated by.
+	// The prefix, namespace, and uid are never truncated.
+	nameOver := len(s) - pubsubMax
+	nameMax := len(n) - nameOver
+
+	return fmt.Sprintf("%s_%s_%s_%s", prefix, ns, hashedTruncate(n, nameMax), uid)
+}
+
+// hashedTruncate truncates a string longer than l by replacing the end of the
+// string with a hash of its contents. Panics if l is less than the length of
+// the hash value (currently 32 characters).
+func hashedTruncate(s string, max int) string {
```

Comment:
Why do we use md5 and why not just simply truncate?

Reply:
Hashing the full string ensures that the pubsub resource names are unique for all kubernetes resource names. Just truncating means that two long broker names with the same first 200 characters map to the same topic name.

Reply:
I think we also have the uid of the broker as the topic name. Is that not enough?

Reply:
That's a good point :) I'll try to rework this a bit.

/hold